// formiojs/components/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _NestedComponent = require('./nested/NestedComponent');

var _NestedComponent2 = _interopRequireDefault(_NestedComponent);

var _Address = require('./address/Address');

var _Address2 = _interopRequireDefault(_Address);

var _Base = require('./base/Base');

var _Base2 = _interopRequireDefault(_Base);

var _Content = require('./content/Content');

var _Content2 = _interopRequireDefault(_Content);

var _Container = require('./container/Container');

var _Container2 = _interopRequireDefault(_Container);

var _DataGrid = require('./datagrid/DataGrid');

var _DataGrid2 = _interopRequireDefault(_DataGrid);

var _DateTime = require('./datetime/DateTime');

var _DateTime2 = _interopRequireDefault(_DateTime);

var _Day = require('./day/Day');

var _Day2 = _interopRequireDefault(_Day);

var _EditGrid = require('./editgrid/EditGrid');

var _EditGrid2 = _interopRequireDefault(_EditGrid);

var _HTML = require('./html/HTML');

var _HTML2 = _interopRequireDefault(_HTML);

var _Hidden = require('./hidden/Hidden');

var _Hidden2 = _interopRequireDefault(_Hidden);

var _Form = require('./form/Form');

var _Form2 = _interopRequireDefault(_Form);

var _TextField = require('./textfield/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _PhoneNumber = require('./phonenumber/PhoneNumber');

var _PhoneNumber2 = _interopRequireDefault(_PhoneNumber);

var _Email = require('./email/Email');

var _Email2 = _interopRequireDefault(_Email);

var _Time = require('./time/Time');

var _Time2 = _interopRequireDefault(_Time);

var _Checkbox = require('./checkbox/Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _Currency = require('./currency/Currency');

var _Currency2 = _interopRequireDefault(_Currency);

var _Fieldset = require('./fieldset/Fieldset');

var _Fieldset2 = _interopRequireDefault(_Fieldset);

var _Signature = require('./signature/Signature');

var _Signature2 = _interopRequireDefault(_Signature);

var _Select = require('./select/Select');

var _Select2 = _interopRequireDefault(_Select);

var _Resource = require('./resource/Resource');

var _Resource2 = _interopRequireDefault(_Resource);

var _TextArea = require('./textarea/TextArea');

var _TextArea2 = _interopRequireDefault(_TextArea);

var _Tags = require('./tags/Tags');

var _Tags2 = _interopRequireDefault(_Tags);

var _Button = require('./button/Button');

var _Button2 = _interopRequireDefault(_Button);

var _Number = require('./number/Number');

var _Number2 = _interopRequireDefault(_Number);

var _Password = require('./password/Password');

var _Password2 = _interopRequireDefault(_Password);

var _Panel = require('./panel/Panel');

var _Panel2 = _interopRequireDefault(_Panel);

var _Column = require('./columns/Column');

var _Column2 = _interopRequireDefault(_Column);

var _Columns = require('./columns/Columns');

var _Columns2 = _interopRequireDefault(_Columns);

var _Table = require('./table/Table');

var _Table2 = _interopRequireDefault(_Table);

var _Unknown = require('./unknown/Unknown');

var _Unknown2 = _interopRequireDefault(_Unknown);

var _Radio = require('./radio/Radio');

var _Radio2 = _interopRequireDefault(_Radio);

var _SelectBoxes = require('./selectboxes/SelectBoxes');

var _SelectBoxes2 = _interopRequireDefault(_SelectBoxes);

var _Survey = require('./survey/Survey');

var _Survey2 = _interopRequireDefault(_Survey);

var _Well = require('./well/Well');

var _Well2 = _interopRequireDefault(_Well);

var _Location = require('./location/Location');

var _Location2 = _interopRequireDefault(_Location);

var _File = require('./file/File');

var _File2 = _interopRequireDefault(_File);

var _Tabs = require('./tabs/Tabs');

var _Tabs2 = _interopRequireDefault(_Tabs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  nested: _NestedComponent2.default,
  address: _Address2.default,
  base: _Base2.default,
  content: _Content2.default,
  container: _Container2.default,
  datagrid: _DataGrid2.default,
  datetime: _DateTime2.default,
  day: _Day2.default,
  htmlelement: _HTML2.default,
  hidden: _Hidden2.default,
  editgrid: _EditGrid2.default,
  form: _Form2.default,
  textfield: _TextField2.default,
  phoneNumber: _PhoneNumber2.default,
  email: _Email2.default,
  time: _Time2.default,
  checkbox: _Checkbox2.default,
  currency: _Currency2.default,
  fieldset: _Fieldset2.default,
  signature: _Signature2.default,
  select: _Select2.default,
  resource: _Resource2.default,
  textarea: _TextArea2.default,
  tags: _Tags2.default,
  button: _Button2.default,
  number: _Number2.default,
  password: _Password2.default,
  panel: _Panel2.default,
  tabs: _Tabs2.default,
  column: _Column2.default,
  columns: _Columns2.default,
  table: _Table2.default,
  unknown: _Unknown2.default,
  radio: _Radio2.default,
  selectboxes: _SelectBoxes2.default,
  survey: _Survey2.default,
  well: _Well2.default,
  location: _Location2.default,
  file: _File2.default
};
